packages/narwallets/src/lib/narwallets.ts

```diff
@@ -0,0 +1,274 @@
+  params: Array<SignAndSendTransactionParams>
+): Promise<
+  Array<providers.FinalExecutionOutcome>
+> => {
+  return sendToNarwallets(NARWALLETS_CODES.SIGN_AND_SEND_TRANSACTIONS, false, params);
+};
+
+const sendToNarwallets = (
+  code: string,
+  withTimeout = false,
+  params?: any
+): Promise<any> => {
+  const promise = new Promise<any>((resolve, reject) => {
+    id++;
+    let promiseTimeout;
+    if (withTimeout) {
+      promiseTimeout = setTimeout(() => {
+        return reject(Error("timeout"));
+      }, 2000);
+    }
+    pendingPromises.push({
+      id,
+      code,
+      resolve,
+      reject,
+      timeout: promiseTimeout,
+    });
+
+    window.postMessage({
+      id,
+      src: "ws",
+      type: "nw",
+      code,
+      dest: "ext",
+      params,
+    });
+  });
+  return promise;
+};
+
+function findPendingPromiseById(id: number): PendingPromises | undefined {
+  return pendingPromises.filter((c) => c.id == id)[0];
+}
+
+function removePendingPromise(callback: PendingPromises) {
+  const index = pendingPromises.indexOf(callback);
+  if (index > -1) {
+    // only splice array when item is found
+    pendingPromises.splice(index, 1); // 2nd parameter means remove one item only
+  }
+}
+
+const setupNarwalletsState = (): void => {
+
+  // receive response from the extension content_script
+  window.addEventListener("message", (event) => {
+    if (event.source != window) {
+      return;
+    }
+    const { data } = event
+    // msg should be directed to the page (response from the extension, relayed from the content script)
+    if (!data || data.dest !== "page") {
+      return
+    }
+
+    if (data.id && data.type == "nw") {
+      const pendingPromise = findPendingPromiseById(data.id);
+      if (pendingPromise) {
+        removePendingPromise(pendingPromise);
+        if (pendingPromise.timeout) {
+          clearTimeout(pendingPromise.timeout);
+        }
+        if (!data.result) {
+          pendingPromise.reject("result is empty");
+        }
+        else if (data.result.err) {
+          pendingPromise.reject(data.result.err);
+        } else {
+          pendingPromise.resolve(data.result.data);
+        }
+      }
+    }
+  });
+};
+
+const Narwallets: WalletBehaviourFactory<InjectedWallet> = async ({
+  options,
+  metadata,
+  store,
+  provider,
+  emitter,
+  logger,
+}) => {
+
+  const signOut = async () => {
+    if (!(await isSignedIn())) {
+      return;
+    }
+
+    const res = await sendToNarwallets("sign-out");
+    // const res = await _state.wallet.signOut();
+    if (res === true) {
+      return;
+    }
+
+    const error = new Error(
+      typeof res.error === "string" ? res.error : res.error.type
+    );
+
+    // Prevent signing out by throwing.
+    if (error.message === "User reject") {
+      throw error;
+    }
+
+    // Continue signing out but log the issue.
+    logger.log("Failed to sign out");
+    logger.error(error);
+  };
+
+  return {
+    async signIn() {
```

Comment:
What does **signIn** into Narwallets mean?

From what I understand Narwallets does not accept a **contractId** during **signIn** to add a FunctionCall Access Key in the chain for the accountId you're accepting the signIn.

Reply:
In this case, signIn is to unlock the wallet. That would be to request the user to input their password.

Reply:
Looks better now, but it blocks the UI when the wallet is locked and you relaod the dApp.